minio/minio
# s3: Allow virtual style endpoints

Fixes: #7696

This change will allow users without ListAllMyBuckets permission
to be able to perform operations allowed as per IAM policy


When user enters virtual style endpoint of the form https://bucket.s3.amazonaws.com, use this bucket for 
all api operations. This would help IAM users with restricted policies to perform allowed operations.

## Description


## Motivation and Context


See referenced issue
## Regression
 No


## How Has This Been Tested?



1. Set up an IAM user with GetObject, ListObject and PutObject permissions  for a bucket
2. Start s3 gateway with `minio gateway s3 https://bucket.s3.amazonaws.com`
3, gateway should come up without errors
4. mc config host add should succeed
5. All operations for which permissions have been granted in step1 should succeed.

## Types of changes


## Checklist:


- [x All new and existing tests passed.
